File: src/BooksReader.Core/Services/ICRUDService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using BooksReader.Core.Models;

namespace BooksReader.Core.Services
{
	public interface ICRUDService<T>  where T : IIdentified
	{
		T Add(T item);
		T Edit(T item);
		IQueryable<T> Get();
		T Get(Guid id);
		T Delete(Guid id);

		Task<T> AddAsync(T item);
		Task<IEnumerable<T>> GetAsync();
		Task<T> GetAsync(Guid id);
	}
}

File: src/BooksReader.Web/ClientApp/src/app/modules/core/services/interfaces/ICRUDService.ts
import { IIdentifiable, OperationResult, StandardFilters } from '@br/core/models';
import { Observable } from 'rxjs';

export interface ICrudService<T extends IIdentifiable>  {
    add(item: T): Observable<OperationResult<T>>;
    get(id: string): Observable<OperationResult<T>>;
    list(filters: StandardFilters): Observable<OperationResult<T[]>>;
    update(item: T): Observable<OperationResult<T>>;
    delete(id: string): Observable<OperationResult<T>>;
}
